### PR TITLE
Adds middleware notifier

### PR DESCRIPTION
**Pull request by williamthome**
### Description

Gives control over how requests are to be processed by a middleware notification.
Handling the request in a module:
```erlang
observe_middleware(#middleware{ on = welformed }, Context, _Context) ->
    io:format("on welformed middleware~n"),
    Context;
observe_middleware(#middleware{ }, Context, _Context) ->
    Context.
```
So custom headers and other stuff can be made during the request.

### Checklist

- [x] documentation updated
- [ ] tests added
- [x] no BC breaks

### CODE REVIEW

File: apps/zotonic_core/include/zotonic_notifications.hrl [hunk 1]
```diff
@@ -1188,6 +1188,14 @@
 }).
 
 
+%% @doc Delegates the request processing.
+%% Type: foldl
+%% Return: modified context
+-record(middleware, {
+    on :: executed | welformed | handled
```
Comment (mworrell) [inline apps/zotonic_core/include/zotonic_notifications.hrl:1195]: Maybe `executed` should be called `init` or `start` ?
That might be clearer.

**Review by mmzeeman — APPROVED** [review]
Nice idea. That allows for some interesting site level customizations. Maybe a little change like @mworrell suggested, because it happens before cowmachine executes the request.

**Comment by vkatsuba** [comment]
Wow PR with number 3000! Zotonic rocks 🤘

File: apps/zotonic_core/include/zotonic_notifications.hrl (hunk 1 above)
Comment (williamthome) [reply apps/zotonic_core/include/zotonic_notifications.hrl:1195]: I'm in doubt about what could be better, but I think `start` would be best. What do you think?

Comment (williamthome) [reply apps/zotonic_core/include/zotonic_notifications.hrl:1195]: Or maybe `request`?

Comment (mworrell) [reply apps/zotonic_core/include/zotonic_notifications.hrl:1195]: `request` sounds good to me.  @mmzeeman @vkatsuba  ok?

Comment (vkatsuba) [reply apps/zotonic_core/include/zotonic_notifications.hrl:1195]: Sounds good.

Comment (williamthome) [reply apps/zotonic_core/include/zotonic_notifications.hrl]: Changed `executed` to `request`.

**Review by vkatsuba — APPROVED** [review]

**Review by mworrell — APPROVED** [review]

**Comment by mworrell** [comment]
Thanks again!

**Pull request merged**